weave/trace_server/clickhouse_trace_server_batched.py

```diff
@@ -1978,6 +1911,329 @@ def completions_create(
+            start_call = _start_call_for_insert_to_ch_insertable_start_call(start)
+            # Insert immediately so that callers can see the call in progress
+            self._insert_call(start_call)
+
+        # --- Build the underlying chunk iterator
+        chunk_iter = lite_llm_completion_stream(
+            api_key=api_key or "",
+            inputs=req.inputs,
+            provider=provider,
+            base_url=base_url,
+            extra_headers=extra_headers,
+            return_type=return_type,
+        )
+
+        # If tracking not requested just return chunks directly
+        if not req.track_llm_call or start_call is None:
+            return chunk_iter
+
+        # Otherwise, wrap the iterator with tracking
+        return self._create_tracked_stream_wrapper(
+            chunk_iter, start_call, model_name, req.project_id
+        )
+
+    def _update_metadata_from_chunk(
+        self, chunk: dict[str, Any], aggregated_metadata: dict[str, Any]
+    ) -> None:
+        """Update aggregated metadata from a chunk."""
+        metadata_fields = [
+            "id",
+            "created",
+            "model",
+            "system_fingerprint",
+            "service_tier",
+            "usage",
+        ]
+
+        for field in metadata_fields:
+            if field in chunk and field not in aggregated_metadata:
+                if field == "service_tier":
+                    aggregated_metadata[field] = chunk.get(field, "default")
+                else:
+                    aggregated_metadata[field] = chunk[field]
+
+    def _process_tool_call_delta(
+        self, tool_call_delta: list, tool_calls: list[dict[str, Any]]
+    ) -> None:
+        """Process tool call delta and update tool_calls list."""
+        for tool_call in tool_call_delta:
+            tool_call_index = tool_call.get("index", 0)
+
+            # Ensure we have enough tool calls in our list
+            while len(tool_calls) <= tool_call_index:
+                tool_calls.append(
+                    {
+                        "id": None,
+                        "type": "function",
+                        "function": {"name": "", "arguments": ""},
+                    }
+                )
+
+            existing_tool_call = tool_calls[tool_call_index]
+
+            # Update existing tool call fields
+            if tool_call.get("id"):
+                existing_tool_call["id"] = tool_call["id"]
+            if tool_call.get("type"):
+                existing_tool_call["type"] = tool_call["type"]
+
+            if "function" in tool_call:
+                function_data = tool_call["function"]
+                if function_data.get("name"):
+                    existing_tool_call["function"]["name"] = function_data["name"]
+                if "arguments" in function_data:
+                    existing_tool_call["function"]["arguments"] += function_data[
+                        "arguments"
+                    ]
+
+    def _clean_tool_calls(
```

Comment:
This function does not need to be an instance method. It doesn't need `self`. Can you define as standalone function? Perhaps in a dedicated module to keep the code well organized?